nifi-nar-bundles/nifi-extension-utils/nifi-hadoop-utils/src/main/java/org/apache/nifi/processors/hadoop/AbstractPutHDFS.java

```diff
@@ -0,0 +1,366 @@
+                    getLogger().error("Failed to write to HDFS due to {}", new Object[]{t});
+                    session.transfer(session.penalize(putFlowFile), getFailureRelationship());
+                    context.yield();
+                }
+
+                return null;
+            }
+        });
+    }
+
+    /**
+     * Returns with the expected block size. Note: this might be overwritten by implementations.
+     */
+    protected long getBlockSize(final ProcessContext context, final ProcessSession session, final FlowFile flowFile) {
+        final String dirValue = context.getProperty(DIRECTORY).evaluateAttributeExpressions(flowFile).getValue();
+        final Path configuredRootDirPath = new Path(dirValue);
+
+        try {
+            return getFileSystem().getFileStatus(configuredRootDirPath).getBlockSize();
+        } catch (final IOException e) {
+            getLogger().warn("Error happened during acquiring status for {}. Determining the block size using default value.", configuredRootDirPath, e);
+            return getFileSystem().getDefaultBlockSize(configuredRootDirPath);
+        }
+    }
+
+    /**
+     * Returns with the expected buffer size. Note: this might be overwritten by implementations.
+     */
+    protected int getBufferSize(final ProcessContext context, final ProcessSession session, final FlowFile flowFile) {
+        return getConfiguration().getInt(BUFFER_SIZE_KEY, BUFFER_SIZE_DEFAULT);
+    }
+
+    /**
+     * Returns with the expected replication factor. Note: this might be overwritten by implementations.
+     */
+    protected short getReplication(final ProcessContext context, final ProcessSession session, final FlowFile flowFile) {
+        final String dirValue = context.getProperty(DIRECTORY).evaluateAttributeExpressions(flowFile).getValue();
+        final Path configuredRootDirPath = new Path(dirValue);
+        return getFileSystem().getDefaultReplication(configuredRootDirPath);
+    }
+
+    /**
+     * Returns if file system should ignore location. Note: this might be overwritten by implementations.
+     */
+    protected boolean shouldIgnoreLocality(final ProcessContext context, final ProcessSession session) {
+        return false;
+    }
+
+    /**
+     * I returns a non-null value, changes the owner of the uploaded file to this value after it is written. This only
+     * works if NiFi is running as a user that has privilege to change owner. Note: this might be overwritten by implementations.
+     */
+    protected String getOwner(final ProcessContext context, final FlowFile flowFile) {
+        return null;
+    }
+
+    /**
+     * I returns a non-null value, changes the group of the uploaded file to this value after it is written. This only
+     * works if NiFi is running as a user that has privilege to change group. Note: this might be overwritten by implementations.
+     */
+    protected String getGroup(final ProcessContext context, final FlowFile flowFile) {
```

Comment:
These methods all have 'default' implementations but are then overridden in the existing concrete implementation. Probably makes more sense to just make them abstract?